tests/integration/test_pip_packages.py

```diff
@@ -0,0 +1,146 @@
+
+    # Download and extract source tarball
+    source_name = tmpdir.join(f"download_{str(completed_response.id)}")
+    file_name_tar = tmpdir.join(f"download_{str(completed_response.id)}.tar.gz")
+    resp = client.download_bundle(completed_response.id, file_name_tar)
+    assert resp.status == 200
+    assert tarfile.is_tarfile(file_name_tar)
+    with tarfile.open(file_name_tar, "r:gz") as tar:
+        tar.extractall(source_name)
+
+    expected_file_urls = test_env["pip_packages"]["without_deps"]["expected_files"]
+
+    # Check that the source tarball includes the application source code under the app directory.
+    assert_expected_files(path.join(source_name, "app"), expected_file_urls)
+    # Check that the source tarball includes an empty deps directory.
+    assert_expected_files(path.join(source_name, "deps"))
+
+    purl = test_env["pip_packages"]["without_deps"]["purl"]
+    image_contents = [{"dependencies": [], "purl": purl, "sources": []}]
+    assert_pip_content_manifest(client, completed_response.id, image_contents)
+
+
+def assert_packages_from_response(response_data, expected_types):
+    """
+    Check amount and type of packages in the response data.
+
+    :param dict response_data: response data from the Cachito request
+    :param list expected_types: expected types of packages
+    """
+    assert len(response_data["packages"]) == len(expected_types)
+    for idx in range(len(expected_types)):
+        assert response_data["packages"][idx]["type"] == expected_types[idx]
+
+
+def assert_dependencies_from_response(response_data, expected_deps=None):
+    """
+    Check that dependencies are correctly listed in the response data.
+
+    :param dict response_data: response data from the Cachito request
+    :param list expected_deps: expected package dependencies
+    """
+    if expected_deps is None:
+        expected_deps = []
+
+    assert set(response_data["dependencies"]) == set(expected_deps)
+
+    packages_deps = []
+    for package in response_data["packages"]:
+        packages_deps += package["dependencies"]
+    assert set(packages_deps) == set(expected_deps)
+
+
+def assert_expected_files(source_path, expected_file_urls=None):
+    """
+    Check that the source path includes expected files.
+
+    :param str source_path: local path for checking
+    :param dict expected_file_urls: {"relative_path/file_name": "URL", ...}
+    """
+    if expected_file_urls is None:
+        expected_file_urls = {}
+    assert path.exists(source_path) and path.isdir(source_path)
+    assert len(listdir(source_path)) == len(expected_file_urls)
+    files = []
+    # Go through all files in source_code_path and it's subdirectories
+    for root, _, source_files in walk(source_path):
+        for file_name in source_files:
+            # Get path to file in the project
+            absolute_file_path = path.join(root, file_name)
+            relative_file_path = absolute_file_path.split(source_path + "/")[1]
```

Comment:
You could use `path.relpath(absolute_file_path, start=source_path)` instead

Reply:
Done, thank you a lot!